### ISSUE TITLE
rename FloatMode.Optimized to FloatMode.Mathematical

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#6634]
from https://ziglang.org/documentation/master/#setFloatMode:

 `FloatMode.Optimized` - Floating point operations may do all of the following:

   * Assume the arguments and result are not NaN. Optimizations are required to retain defined behavior over NaNs, but the value of the result is undefined.
   *  Assume the arguments and result are not +/-Inf. Optimizations are required to retain defined behavior over +/-Inf, but the value of the result is undefined.
   *   Treat the sign of a zero argument or result as insignificant.
   *   Use the reciprocal of an argument rather than perform division.
   *   Perform floating-point contraction (e.g. fusing a multiply followed by an addition into a fused multiply-and-add).
   *   Perform algebraically equivalent transformations that may change results in floating point (e.g. reassociate).

The idea behind these transformations is to remove the possibility of things that are "not a number" and pretend that we are actually talking about real numbers only. So better to call it something that reflects that rather than call it "fast" or "optimized".

**Comment by gereeter** [comment]
This seems misleading; I'd expect a "mathematical" mode to act predictably and follow mathematical laws, or maybe improve numerical accuracy, but that's not what the fast math flags do. They instead break more laws (technically you can't even rely on `x == x` to be true, for example) and can sometimes worsen numerical accuracy in search of performance.

Perhaps this would make sense if `@setFloatMode` was renamed to `@setFloatAssumptions`?

**Comment by tadeokondrak** [comment]
The last three points seem like they can be important even if you don't have any NaNs.

**Comment by ityonemo** [comment]
Hm.  Having worked extensively on IEEE floating point and alternatives, I don't particularly like calling it "mathematical" (even as much as I hate the non-mathey choices of IEEE 754), because +/- inf are certainly mathematical concepts that 754 tried to incorporate.

At best you might have a claim to "arithmetic", but optimizable, I would think is okay.  Is there a distinction here from "fastmath"?

Also with the "insignificant sign on zero", be sure to double check what your impls are doing with sqrt(-0) because some might throw nan, some might return -0.

**Comment by Sobeston** [comment]
I find the term mathematical unhelpful, I read it and don't know what it actually means. It doesn't convey anything to me. Something like fast, optimised, fast-math, etc conveys to me that things may be slightly approximate and non-standard for the sake of speed. Sure that isn't the whole story, but I think the current name is more telling.

**Comment by tgschultz** [comment]
I agree with others that calling it `Mathematical` is at best misleading. Basically what `FloatMode.Optimized` does is relax the well-defined behavior of IEEE floats right? How about `FloatMode.Relaxed`, which pairs nicely with `FloatMode.Strict`?

**Comment by Hadron67** [comment]
Well from a mathematical point of view, Inf and NaN do have meanings, they both can be regarded as [singularitiy](https://en.wikipedia.org/wiki/Singularity_(mathematics)), their arithmetic behaviours and function values also have mathematical backgrounds:

Inf represents a [pole](https://en.wikipedia.org/wiki/Zeros_and_poles) or *infinite discontinuity*, where the value definitely goes infinity, like 1 / 0 = Inf, exp(+Inf) = +Inf, exp(-Inf) = 0, log(+0) = -Inf.

NaNs arise mainly in 3 ways:
* a number outside of the real number field, like sqrt(-1);
* an [essential singularity](https://en.wikipedia.org/wiki/Essential_singularity), where the value neither converges to some real number nor goes infinity. The most famous example of this is sin(Inf) = NaN.
* [indeterminate forms](https://en.wikipedia.org/wiki/Indeterminate_form): `Inf / Inf`, `0 / 0`, `1 ** Inf`, `0 ** 0`, `Inf * 0`, `Inf ** 0` (`**` means exponentiation), since their values depends on how the functions approach 0 or Inf, their values can be any number and thus all evaluate to NaN.

Summing up, NaN represents *a number that can be any number or no number*, so any operation on NaN would also give NaN.

Also as mentioned in the above comments, the sign of zero is also mathematically significant, as some function might have discontinuity at zero: exp(1 / x) gives Inf for +0, and 0 for -0.

So in this way, I'd expect `FloatMode.Mathematical` gives floating point numbers a mathematical sensible behaviour as shown above. Although in practice these edge cases are of little significance, the name `Mathematical` is rather misleading and something like `Optimized` or `Fast` can be much better.

**Comment by andrewrk** [comment]
closing in favor of #23173
